Flink anomaly detection tutorial (#1704)
aiven/devportal: docs/tutorials/anomaly-detection.rst

Streaming anomaly detection with Apache Flink®, Apache Kafka® and PostgreSQL®
==============================================================================================

.. Note::

>     This tutorial doesn't assume any existing Apache Kafka®, PostgreSQL® or Apache Flink® knowledge

Author: true, removed

What you will learn
---------------------------

The tutorial provides an example of a streaming anomaly detection system, by following it you'll learn:

* How to create a fake streaming dataset
* How to create and use Apache Kafka® for data streaming
* How to create and use PostgreSQL® to store threshold data
* How to create and use Apache Flink® to define streaming data pipelines
* How to push the outcome of the anomaly detection system as a Slack notification

What are you going to build
---------------------------

In this tutorial you will build a streaming anomaly detection system based on IoT sensor readings recording the CPU usage percentage for various devices. After setting up the various components and enabling the data flow, you'll then create a basic filtering pipeline to separate the usage values surpassing a fixed threshold (80%). 

.. mermaid::

    graph LR;

>         id1(IoT device)-- sensor reading -->id2(usage > 80%?);

Reviewer: Can we have a small but meaningful example behind what sensor reading we're looking for?
It will be easier to follow, if it is something relatable. For example, you can suggest that the sensors measure the temperature or humidity across the building, or something else. And then follow that story.

.. mermaid::

    graph LR;

        id1(IoT device)-- sensor reading -->id2(average 30 seconds);
        id2-- average reading -->id4(over threshold?);
        id4-- yes --> id5(notification);
        id3(threshold table)-- thresholds --> id4;

.. Note::

    Even if the sample dataset and the data pipeline examples might seem basic, they offer a wide coverage on the integration and transformation options available with Apache Flink that can be applied to other, more complex, scenarios. 

The tutorial includes:

* Apache Flink for data transformation
* Apache Kafka for data streaming
* PostgreSQL® for data storage/query
* Slack as notification system

.. Tip::

    All the tools listed above are fully open source, and can therefore be installed and used locally. You're going to use the comparable Aiven services to avoid any network and integration complexities. 

Architecture overview
---------------------

> The tutorial showcases how to create an Apache Kafka® source topic that provides a stream of IoT metrics data, a PostgreSQL® database that contains data on the alerting thresholds, and an Apache Flink® service that combines these two services and pushes the filtered data to a separate Apache Kafka® topic or a Slack channel.

Reviewer: This is similarly in a very dry language. Maybe it is the goal, but maybe we want to make it friendlier, like blog style? I suspect tutorials are more like blogs, rather than docs.

* An :doc:`Aiven for Apache Flink®</docs/products/flink>` named ``demo-flink`` for streaming data transformation
* An :doc:`Aiven for PostgreSQL®</docs/products/postgresql>` named ``demo-postgresql`` for data storage and query

Create an Aiven for Apache Kafka® service
'''''''''''''''''''''''''''''''''''''''''''''

The :doc:`Aiven for Apache Kafka </docs/products/kafka>` service is responsible to receive the inbound stream of IoT sensor readings. To create the service you need to:

1. Log in to the `Aiven web console <https://console.aiven.io/>`_.
2. On the *Services* page, click **Create a new service**.

   This opens a new page with the available service options.

   .. image:: /images/platform/concepts/console_create_service.png
      :alt: Aiven Console view for creating a new service

3. Select **Apache Kafka®**.

4. Select the cloud provider and region that you want to run your service on.

5. Select `business-4` as service plan.

5. Enter ``demo-kafka`` as name for your service.

6. Click **Create Service** under the summary on the right side of the console

Customise the Aiven for Apache Kafka service
''''''''''''''''''''''''''''''''''''''''''''

> After creating the service, you'll be redirected to the service details page. You can now customise the service to enable the needed components in the *Overview* tab:

Reviewer: I'd say it something like this -

"Now that your service is created, you need to customise it a bit.

On the Overview tab of your freshly created service, you'll see a bunch of toggles and properties.  Change these two:"

Reviewer: Or maybe this is a bit too conversational tone? :D Leaving these ideas, but up to you if you want to use them, Francesco!

1. **Kafka REST API (Karapace)** > **Enable**

   .. Note:: 

>     The **Kafka REST API** enables you to manage Apache Kafka via REST APIs and also to view the data in your Apache Kafka® topics.

Reviewer: This is quote from the docs, but why we need it for this tutorial? If you have to use less formal lanuage?

2. **Advanced configuration** > **Add configuration option** > ``kafka.auto_create_topics_enable``, switch the setting on and then click **Save advanced configuration**

   .. Note:: 
    The ``kafka.auto_create_topics_enable`` setting allows you to create new Apache Kafka® topics as you configure your Apache Flink® data tables, so that you do not need to create the topics in advance.

Create an Aiven for PostgreSQL® service

> '''''''''''''''''''''''''''''''''''''''''

Reviewer: Maybe a sentence about the goal of this PostgreSQL service? It was on the diagram, but some easy to read one sentence can be helpful here too

1. Log in to the `Aiven web console <https://console.aiven.io/>`_.
2. On the *Services* page, click **Create a new service**.

3. Select **PostgreSQL®**.

4. Select the cloud provider and region that you want to run your service on.

5. Select `hobbyist` as service plan.

5. Enter ``demo-posgresql`` as name for your service.

6. Click **Create Service** under the summary on the right side of the console

Create an Aiven for Apache Flink service
'''''''''''''''''''''''''''''''''''''''''

You will use :doc:`Apache Flink </docs/products/flink>` to define the streaming data pipelines. To create the service:

1. Log in to the `Aiven web console <https://console.aiven.io/>`_.
2. On the *Services* page, click **Create a new service**.

3. Select **Apache Flink®**.

4. Select the cloud provider and region that you want to run your service on.

5. Select `business-4` as service plan.

5. Enter ``demo-flink`` as name for your service.

6. Click **Create Service** under the summary on the right side of the console

Customise the Aiven for Apache Flink service
''''''''''''''''''''''''''''''''''''''''''''

> After creating the service, you'll be redirected to the service details page. You can now customise the service to enable the needed integrations to the Aiven for Apache Kafka and Aiven for PostgreSQL services in the *Overview* tab:

Reviewer: This is the first time you mention integrations between Aiven services.
Maybe this section is not "Customise the Aiven for Apache Flink service", but "Connect the building blocks"

Architecture overview
---------------------

In this tutorial you'll build a streaming data pipeline. The sources will be an Apache Kafka topic, containing the fake stream of IoT metrics data, and a PostgreSQL® database that containing the alerting thresholds. Then, an Apache Flink® service, will combine the data, apply some transformation SQL to find the anomalies, and push the result to a separate Apache Kafka® topic or a Slack channel for team notification.

.. mermaid::

    graph TD;

        id1(Apache Kafka)-- IoT metrics stream -->id3(Apache Flink);
        id2(PostgreSQL)-- alerting threshold data -->id3;
        id3-- filtered/aggregated data -->id1;
        id3-- filtered data -->id7(Slack);

Prerequisites
-------------

The tutorial uses Aiven services, therefore you'll need a valid `Aiven account <https://console.aiven.io/signup>`_. On top of the Aiven account, you will also need the following three items:

* **Docker**, needed for the `fake data generator for Apache Kafka <https://github.com/aiven/fake-data-producer-for-apache-kafka-docker>`_. Check out the `related installation instructions <https://docs.docker.com/engine/install/>`_.

> * **Slack App and Token**: the data pipeline output is a notifications to a Slack channel, check out the needed steps to `set up a Slack app and retrieve the Slack authentication token <https://github.com/aiven/slack-connector-for-apache-flink>`_.

Reviewer: The only part that made me stuck is understanding Slack connector :D and the info in github isn't too helpful for those with 0 knowledge.... some more hints would be highly appreciated :)

Author: I believe this will make the tutorial even longer. I tried making the instruction in GH a bit easier to follow with a PR recently, since most of the work needs to be performed in Slack, I would leave this out of scope for the tutorial

Create an Aiven for Apache Kafka® service
'''''''''''''''''''''''''''''''''''''''''''''

The :doc:`Aiven for Apache Kafka </docs/products/kafka>` service is responsible to receive the inbound stream of IoT sensor readings. You can create the service with the following steps:

1. Log in to the `Aiven web console <https://console.aiven.io/>`_.
2. On the *Services* page, click **Create a new service**.

   This opens a new page with the available service options.

   .. image:: /images/platform/concepts/console_create_service.png
      :alt: Aiven Console view for creating a new service

3. Select **Apache Kafka®**.

4. Select the cloud provider and region that you want to run your service on.

5. Select `business-4` as service plan.

5. Enter ``demo-kafka`` as name for your service.

6. Click **Create Service** under the summary on the right side of the console

Customise the Aiven for Apache Kafka service
''''''''''''''''''''''''''''''''''''''''''''

Now that your service is created, you need to customise its functionality. In the **Overview** tab of your freshly created service, you'll see a bunch of toggles and properties. Change these two:

> 1. **Kafka REST API (Karapace)** > **Enable**

Reviewer: I'd remove the Note and add why we need it as a part of a story

   .. Note::

    The **Kafka REST API** allows you to manage and query Apache Kafka via REST APIs. You'll use it to inspect the data in Apache Kafka from the Aiven Console.

2. **Advanced configuration** > **Add configuration option** > ``kafka.auto_create_topics_enable``, switch the setting on and then click **Save advanced configuration**

   .. Note::

>     The ``kafka.auto_create_topics_enable`` setting allows you to create new Apache Kafka® topics on the fly while pushing a first record. It avoids needing to create a topic in advance. To read more about the setting, check the :doc:`dedicated documentation </docs/products/kafka/howto/create-topics-automatically>`.

Reviewer: Because this is a tutorial for the sake of a story flow I'd remove the note and refrase this section explaining why we need to set this setting

3. Click **Integrate**.
4. Click the **+** icon under *Data Flow*.
5. Check the **Aiven for PostgreSQL** checkbox in the `Aiven Data Services` section.
6. Select **Aiven for PostgreSQL®** and then select the ``demo-postgresql`` service.
7. Click **Integrate**.

Once the above steps are completed, your **Data Flow** section should be similar to the below

.. image:: /images/tutorials/anomaly-detection/flink-integrations-done.png
      :alt: Aiven for Apache Flink Overview tab, showing the Integrations to Aiven for Apache Kafka and Aiven for PostgreSQL

Set up the IoT metrics streaming dataset
----------------------------------------

Now that the plumbing of all the components is sorted, it's time for you to create a continuous stream of fake IoT data that will land in an Aiven for Apache Kafka topic. There are various ways to generate fake data, for the tutorial you'll use the `Dockerized fake data producer for Aiven for Apache Kafka® <https://github.com/aiven/fake-data-producer-for-apache-kafka-docker>`_ allowing you to generate a continuous flow of data with a minimal setup.

Create an Aiven authentication token
''''''''''''''''''''''''''''''''''''

> To generate the fake streaming IoT metric data against to the Aiven for Apache Kafka topic you'll use the `Dockerized fake data producer for Aiven for Apache Kafka® <https://github.com/aiven/fake-data-producer-for-apache-kafka-docker>`_ which requires an Aiven authentication token to fetch all the Apache Kafka connection parameters. 

Reviewer: ```suggestion
To generate the fake streaming IoT metric data against to the Aiven for Apache Kafka topic use the `Dockerized fake data producer for Aiven for Apache Kafka® <https://github.com/aiven/fake-data-producer-for-apache-kafka-docker>`_ which requires an Aiven authentication token to fetch all the Apache Kafka connection parameters. 
```

> To generate the fake streaming IoT metric data against to the Aiven for Apache Kafka topic you'll use the `Dockerized fake data producer for Aiven for Apache Kafka® <https://github.com/aiven/fake-data-producer-for-apache-kafka-docker>`_ which requires an Aiven authentication token to fetch all the Apache Kafka connection parameters. 

Reviewer: you kinda repeat here what you said above....

   .. image:: /images/tutorials/anomaly-detection/filtering-preview.png
      :alt: The Apache Flink data transformation with SQL preview of the data

9. Click **Save and deploy later**
10. Click **Create deployment**. 
11. Accept the default deployment parameters and click on **Deploy without a savepoint**

    .. image:: /images/tutorials/anomaly-detection/filtering-application-deployment.png
        :alt: Detail of the new deployment screen showing the default version, savepoint and parallelism parameters

12. The new application deployment status will show **Initializing** and then **Running: version 1**.

Once the application is running, you should start to see messages indicating hosts with high CPU loads in the ``cpu_load_stats_real_filter`` topic of your ``demo-kafka`` Apache Kafka service.

.. image:: /images/tutorials/anomaly-detection/filtering-topic-preview.png
      :alt: The Apache Flink data transformation with SQL preview of the data

.. Important::
    
    Congratulations! You created your first streaming anomaly detection pipeline!

> Create an advanced anomaly detection pipeline with windowing and threshold lookup

Reviewer: Maybe better to name it "Adding improvements" or "Evolving" (I can't come up with a good name), a name that will indicate that the work will be on top of what is created already, and not a separate new tutorial

Streaming anomaly detection with Apache Flink®, Apache Kafka® and PostgreSQL®
==============================================================================================

What you will learn
---------------------------

Follow this tutorial and you'll learn how to build a streaming anomaly detection system. In particular, we'll cover the following:

* How to create a fake streaming dataset
* How to create and use Apache Kafka for data streaming
* How to create and use PostgreSQL® to store threshold data
* How to create and use Apache Flink® to define streaming data pipelines
* How to push the outcome of the anomaly detection system as a Slack notification

What are you going to build
---------------------------

> Anomaly detection is a way to find unusual or unexpected things in data. It is immensely helpful in a variety of fields, such as fraud detection, network security, quality control and others. By following this tutorial you will build your own streaming anomaly detection system. 

Reviewer 2: I'd expand this with a practical example:

```suggestion
Anomaly detection is a way to find unusual or unexpected things in data. It is immensely helpful in a variety of fields, such as fraud detection, network security, quality control and others. By following this tutorial you will build your own streaming anomaly detection system. 

For example: a payment processor might set up anomaly detection against an e-commerce store if it notices that the store – which sells its items in Indian Rupees and is only configured to sell to the Indian market – is suddenly receiving a high volume of orders from Spain. This behavior could indicate fraud. Another example is that of a domain hosting service implementing a CAPTCHA against an IP address it deems is interacting with one of its domains in rapid succession.
```

        :alt: Detail of the new deployment screen showing the default version, savepoint and parallelism parameters

13. The new application deployment status will show **Initializing** and then **Running: version 1**.

Once the application is running, you should start to see messages indicating hosts with high CPU loads in the ``cpu_load_stats_real_filter`` topic of your ``demo-kafka`` Apache Kafka service.

.. image:: /images/tutorials/anomaly-detection/filtering-topic-preview.png
      :alt: The Apache Flink data transformation with SQL preview of the data

.. Important::
    
    Congratulations! You created your first streaming anomaly detection pipeline! 

    The data is now available in the Apache Kafka topic named ``cpu_load_stats_real_filter``, from there you could either write your own :doc:`Apache Kafka consumer </docs/products/kafka/howto/list-code-samples>` to read the high sensor records or use :doc:`Kafka Connect </docs/products/kafka/kafka-connect>` to sink the data to a wide range of technologies.

Evolve the anomaly detection pipeline with windowing and threshold lookup
---------------------------------------------------------------------------------

> Sending an alert on every IoT measurement above a threshold might cause too much noise, you don't want to receive a notification every time your computer's CPU goes above 90%, but, if that happens continuously for a 30 seconds interval there might be a problem. In the second example, you'll aggregate the CPU load over a configured time using :doc:`windows </docs/products/flink/concepts/windows>` and the :doc:`event time </docs/products/flink/concepts/event-processing-time>`. By averaging the CPU values over a time window you can encompass short term spikes in usage, and flag only anomaly scenarios where the usage is consistently above a pre-defined threshold for a long period of time.

Reviewer 2: ```suggestion
In most production environments, you wouldn't want to send an alert on every measurement above the threshold. Sometimes CPUs spike momentarily, for example, and come back down in. useage milliseconds later. What's really useful to you in production is if a CPU spike is sustained over a certain period of tim.e 

If a CPU useage spike happens continuously for a 30 seconds interval, there might be a problem. In this step, you'll aggregate the CPU load over a configured time using :doc:`windows </docs/products/flink/concepts/windows>` and the :doc:`event time </docs/products/flink/concepts/event-processing-time>`. By averaging the CPU values over a time window you can filter out short term spikes in usage, and flag only anomaly scenarios where the usage is consistently above a pre-defined threshold for a long period of time.
```

> To add a bit of complexity, and mimic a real scenario, it is better to move away from a fixed ``80%`` threshold, and compare the average utilization figures with the different thresholds, set in a reference table (stored in PostgreSQL), for the various IoT devices based on their ``hostname``. Every IoT device is different, and various devices usually have different alerting ranges. The reference table provides an example of variable, device dependant, thresholds. 

Reviewer 2: ```suggestion
To add a bit of complexity, and mimic a real scenario, we'll also move away from a fixed ``80%`` threshold, and compare the average utilization figures with the different thresholds, set in a reference table (stored in PostgreSQL), for the various IoT devices based on their ``hostname``. Every IoT device is different, and various devices usually have different alerting ranges. The reference table provides an example of variable, device dependent, thresholds. 
```
